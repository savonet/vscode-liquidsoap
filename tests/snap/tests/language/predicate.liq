success = ref(true)

def predicate.to_list(n, p)
  list.init(n, fun(_) -> p())
end

def predicate.nth(n)
  i = ref(-1)
  fun () -> begin
    ref.incr(i)
    i() == n
  end
end

def predicate.from(n)
  i = ref(-1)
  fun () -> begin
    ref.incr(i)
    i() >= n
  end
end

def t(x, y)
  if x != y then
    print("Failure: got #{x} instead of #{y}")
    success := false
  end
end

def tl(p, l)
  t(predicate.to_list(4, p), l)
end

def f() =
  tl(predicate.nth(2), [false,false,true,false])
  tl(predicate.from(2), [false,false,true,true])
  tl(predicate.activates(predicate.from(2)), [false,false,true,false])
  tl(predicate.changes(predicate.nth(2)), [false,true,true,false])
  tl(predicate.once(predicate.from(2)), [false,false,true,false])

  p = predicate.signal()
  t(p(), false)
  p.signal()
  t(p(), true)
  t(p(), false)

  if success() then test.pass() else test.fail() end
end

test.check(f)
