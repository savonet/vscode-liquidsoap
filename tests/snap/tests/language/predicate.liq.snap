>success = ref(true)
#^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
>
>def predicate.to_list(n, p)
#^^^ source.liquidsoap keyword.control.liquidsoap
#   ^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
>  list.init(n, fun(_) -> p())
#^^^^^^^^^^^^^^^ source.liquidsoap
#               ^^^ source.liquidsoap keyword.control.liquidsoap
#                  ^^^^^^^^^^^^ source.liquidsoap
>end
#^^^ source.liquidsoap keyword.control.liquidsoap
>
>def predicate.nth(n)
#^^^ source.liquidsoap keyword.control.liquidsoap
#   ^^^^^^^^^^^^^^^^^^ source.liquidsoap
>  i = ref(-1)
#^^^^^^^^^^^ source.liquidsoap
#           ^ source.liquidsoap constant.numeric.integer
#            ^^ source.liquidsoap
>  fun () -> begin
#^^ source.liquidsoap
#  ^^^ source.liquidsoap keyword.control.liquidsoap
#     ^^^^^^^ source.liquidsoap
#            ^^^^^ source.liquidsoap keyword.control.liquidsoap
>    ref.incr(i)
#^^^^^^^^^^^^^^^^ source.liquidsoap
>    i() == n
#^^^^^^^^^^^^^ source.liquidsoap
>  end
#^^ source.liquidsoap
#  ^^^ source.liquidsoap keyword.control.liquidsoap
>end
#^^^ source.liquidsoap keyword.control.liquidsoap
>
>def predicate.from(n)
#^^^ source.liquidsoap keyword.control.liquidsoap
#   ^^^^^^^^^^^^^^^^^^^ source.liquidsoap
>  i = ref(-1)
#^^^^^^^^^^^ source.liquidsoap
#           ^ source.liquidsoap constant.numeric.integer
#            ^^ source.liquidsoap
>  fun () -> begin
#^^ source.liquidsoap
#  ^^^ source.liquidsoap keyword.control.liquidsoap
#     ^^^^^^^ source.liquidsoap
#            ^^^^^ source.liquidsoap keyword.control.liquidsoap
>    ref.incr(i)
#^^^^^^^^^^^^^^^^ source.liquidsoap
>    i() >= n
#^^^^^^^^^^^^^ source.liquidsoap
>  end
#^^ source.liquidsoap
#  ^^^ source.liquidsoap keyword.control.liquidsoap
>end
#^^^ source.liquidsoap keyword.control.liquidsoap
>
>def t(x, y)
#^^^ source.liquidsoap keyword.control.liquidsoap
#   ^^^^^^^^^ source.liquidsoap
>  if x != y then
#^^ source.liquidsoap
#  ^^ source.liquidsoap keyword.control.liquidsoap
#    ^^^^^^^^ source.liquidsoap
#            ^^^^ source.liquidsoap keyword.control.liquidsoap
>    print("Failure: got #{x} instead of #{y}")
#^^^^^^^^^^ source.liquidsoap
#          ^ source.liquidsoap string.quoted.double.liquidsoap
#           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap string.quoted.double.liquidsoap
#                                            ^ source.liquidsoap string.quoted.double.liquidsoap
#                                             ^^ source.liquidsoap
>    success := false
#^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
>  end
#^^ source.liquidsoap
#  ^^^ source.liquidsoap keyword.control.liquidsoap
>end
#^^^ source.liquidsoap keyword.control.liquidsoap
>
>def tl(p, l)
#^^^ source.liquidsoap keyword.control.liquidsoap
#   ^^^^^^^^^^ source.liquidsoap
>  t(predicate.to_list(4, p), l)
#^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
#                      ^ source.liquidsoap constant.numeric.integer
#                       ^^^^^^^^^ source.liquidsoap
>end
#^^^ source.liquidsoap keyword.control.liquidsoap
>
>def f() =
#^^^ source.liquidsoap keyword.control.liquidsoap
#   ^^^^^^^ source.liquidsoap
>  tl(predicate.nth(2), [false,false,true,false])
#^^^^^^^^^^^^^^^^^^^ source.liquidsoap
#                   ^ source.liquidsoap constant.numeric.integer
#                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
>  tl(predicate.from(2), [false,false,true,true])
#^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
#                    ^ source.liquidsoap constant.numeric.integer
#                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
>  tl(predicate.activates(predicate.from(2)), [false,false,true,false])
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
#                                        ^ source.liquidsoap constant.numeric.integer
#                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
>  tl(predicate.changes(predicate.nth(2)), [false,true,true,false])
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
#                                     ^ source.liquidsoap constant.numeric.integer
#                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
>  tl(predicate.once(predicate.from(2)), [false,false,true,false])
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
#                                   ^ source.liquidsoap constant.numeric.integer
#                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
>
>  p = predicate.signal()
#^^^^^^^^^^^^^^^^^^^^^^^^^ source.liquidsoap
>  t(p(), false)
#^^^^^^^^^^^^^^^^ source.liquidsoap
>  p.signal()
#^^^^^^^^^^^^^ source.liquidsoap
>  t(p(), true)
#^^^^^^^^^^^^^^^ source.liquidsoap
>  t(p(), false)
#^^^^^^^^^^^^^^^^ source.liquidsoap
>
>  if success() then test.pass() else test.fail() end
#^^ source.liquidsoap
#  ^^ source.liquidsoap keyword.control.liquidsoap
#    ^^^^^^^^^^^ source.liquidsoap
#               ^^^^ source.liquidsoap keyword.control.liquidsoap
#                   ^^^^^^^^^^^^^ source.liquidsoap
#                                ^^^^ source.liquidsoap keyword.control.liquidsoap
#                                    ^^^^^^^^^^^^^ source.liquidsoap
#                                                 ^^^ source.liquidsoap keyword.control.liquidsoap
>end
#^^^ source.liquidsoap keyword.control.liquidsoap
>
>test.check(f)
#^^^^^^^^^^^^^^ source.liquidsoap
>